server/fishtest/rundb.py

```diff
@@ -477,70 +479,68 @@ def flush_all(self):
                     oldest_entry = cache_entry
             if oldest_entry is not None:
                 oldest_run = oldest_entry["run"]
-                self.scavenge(oldest_run)
-                oldest_entry["last_scavenge_time"] = now
                 oldest_entry["is_changed"] = False
                 oldest_entry["last_sync_time"] = time.time()
                 with self.run_cache_write_lock:
                     self.runs.replace_one({"_id": oldest_run["_id"]}, oldest_run)
-        except Exception as e:
-            print(f"Flush exception: {str(e)}", flush=True)
-        finally:
-            self.run_cache_lock.release()
-
-    def scavenge(self, run):
-        if datetime.now(timezone.utc) < boot_time + timedelta(seconds=300):
-            return False
-        dead_task = False
-        old = datetime.now(timezone.utc) - timedelta(minutes=6)
-        task_id = -1
-        for task in run["tasks"]:
-            task_id += 1
-            if task["active"] and task["last_updated"] < old:
-                self.set_inactive_task(task_id, run)
-                dead_task = True
-                print(
-                    "dead task: run: https://tests.stockfishchess.org/tests/view/{} task_id: {} worker: {}".format(
-                        run["_id"], task_id, worker_name(task["worker_info"])
-                    ),
-                    flush=True,
-                )
-                self.handle_crash_or_time(run, task_id)
-                self.actiondb.dead_task(
-                    username=task["worker_info"]["username"],
-                    run=run,
-                    task_id=task_id,
-                )
-        return dead_task
+
+    def clean_cache(self):
+        now = time.time()
+        with self.run_cache_lock:
+            # We make this a list to be able to change run_cache during iteration
+            for r_id, cache_entry in list(self.run_cache.items()):
+                run = cache_entry["run"]
+                # Presently run["finished"] implies run["cores"]==0 but
+                # this was not always true in the past.
+                if (
+                    not cache_entry["is_changed"]
+                    and (run["cores"] <= 0 or run["finished"])
+                    and cache_entry["last_access_time"] < now - 300
+                ):
+                    del self.run_cache[r_id]
+
+    def scavenge_dead_tasks(self):
+        now = time.time()
+        dead_tasks = []
+        with self.run_cache_lock:
+            for cache_entry in self.run_cache.values():
+                run = cache_entry["run"]
+                if not cache_entry["is_changed"] and not run["finished"]:
+                    for task_id, task in enumerate(run["tasks"]):
+                        if (
+                            task["active"]
+                            and task["last_updated"].timestamp() < now - 360
+                        ):
+                            cache_entry["is_changed"] = True
```

Comment:
I wonder if setting here "is_changed" and then releasing the lock can have an unexpected side effect

Reply:
Code instrumented.
```diff
diff --git a/server/fishtest/rundb.py b/server/fishtest/rundb.py
index 2e9814b2..e2a5bf9f 100644
--- a/server/fishtest/rundb.py
+++ b/server/fishtest/rundb.py
@@ -513,8 +513,12 @@ class RunDb:

     def scavenge_dead_tasks(self):
         now = time.time()
+        print("scavenge_dead_tasks: start", flush=True)
         dead_tasks = []
         with self.run_cache_lock:
+            print("scavenge_dead_tasks: get lock", flush=True)
+            print(f"run_cache count: {len(self.run_cache)}", flush=True)
+            dead_tasks_count = 0
             for cache_entry in self.run_cache.values():
                 run = cache_entry["run"]
                 if not cache_entry["is_changed"] and not run["finished"]:
@@ -525,7 +529,10 @@ class RunDb:
                         ):
                             cache_entry["is_changed"] = True
                             dead_tasks.append((task_id, run))
+                            dead_tasks_count += 1
         # We release the lock to avoid deadlock
+        print("scavenge_dead_tasks: release lock", flush=True)
+        print(f"{dead_tasks_count=}", flush=True)
         for task_id, run in dead_tasks:
             task = run["tasks"][task_id]
             print(
```

Reply:
[dead_task.zip](https://github.com/user-attachments/files/15522575/dead_task.zip)

The instrumented count confirms the number of dead tasks scavenged